### ISSUE TITLE
@!attribute completely broken

### ISSUE THREAD

**Issue by paddor** [lsegal/yard#917]
Maybe this is related to #874.

Consider this code:

```ruby
    # @!attribute [rw] name
    #   Name of the item.
    # @return [String] name of the item
    def name
      ffi_delegate.name.read_string
    end

    def name=(new_name)
      ffi_delegate.set_name(new_name)
    end
```

Good things:

* it does document `name` as an instance attribute, including its type
* it does not document the method `#name` and `#name=` themselves

Problems:

* changing from `rw` to `r` makes no difference
  * it does not mark it as read-only
  * it still doesn't document the method `#name=`
* return type is only recognized if unindented
  * even though documentation of `@!attribute` shows it indented (see [1])
* no source code is made accessible in the documentation
  * even though the documentation of `@!attribute` claims that (see [1])

[[1]](http://www.rubydoc.info/gems/yard/file/docs/Tags.md#attribute)

**Comment by lsegal** [comment]
I wouldn't say "completely broken".

> changing from rw to r makes no difference

It's *not* read-only, you've defined a writer method, so by definition (literally), it's parsed as readwrite. `@!attribute` is primarily meant for scenarios where there is no explicit method declaration, like a DSL method, eval, or native code. If you have direct code, you should be documenting directly on the method.

Basically, you're confusing YARD by using a declaration to fake a method declaration followed by YARD actually parsing a real method declaration. See the docs on directives for more context:

> Directives are similar to meta-data tags in the way they are specified, but they do not add meta-data to the object directly. Instead, they affect the parsing context and objects themselves, allowing a developer to create objects (like methods) outright, rather than simply add text to an existing object.

You don't need to affect parsing in order to add a docstring to #name, you simply need a directive to mark is as an attribute, which can be done as a *separate* comment line:

```ruby
## @!attribute name

# @return [String] name
def name; end

...
```

Or, alternatively, document ONLY the directive, not the method definition itself:

```ruby
# @!attribute name
#   @return [String] name


def name; end
# note multiple spaces of separation
```

Basically, never mix a directive that creates a code object with an explicit code definition of a code object in the same docstring. Otherwise YARD will parse your directive, then see a regular (empty) docstring defined on the same object and overwrite all your hard work with the empty docstring. The last parsed docstring always wins.

> return type is only recognized if unindented

The above explanation applies to this point as well. By attaching the directive to #name's docstring, you're effectively causing YARD to overwrite the docstring with empty text after it parses the directive, wiping out your return tag. If you unindent the tag, it sticks around because it's being parsed after the directive's docstring additions.

It's debatable whether YARD should omit overwriting an existing docstring with an empty docstring, but changing this may break assumptions that other docs might be relying on so it would have to be investigated further. This is something we *could* improve on.

> no source code is made accessible in the documentation

tl;dr no attributes display source code in YARD's default HTML template-- this is not specific to the directive usage.

I think this might be a misunderstanding of the verbiage here. Source *is* added to the YARD database for the code object. In other words, YARD is storing that information internally. The issue here is that the default HTML template that YARD uses will *never* expose source for attributes, whether there is source attached or not. The parser parses it-- the *default* template is omitting it  (custom templating could display it, if needed).

This is an explicit decision in the templating layer due to the fact that the implementation for an attribute is typically irrelevant. Implementation details are usually only needed in documentation if the implementation itself is "doing something"-- in an attribute's case, you are only accessing other properties, and, typically, internal properties that are not relevant to documentation. If you think the implementation for your attribute is important, i.e., it "does something"-- consider the fact that it might not be an "attribute"-- especially if you also believe that it should be considered readonly.

**Comment by paddor** [comment]
Thanks for the detailed response. That clarifies a lot about YARD.

So I assume this means that when you have an attribute whose getter and setter are not as simple as `attr_accessor :foo`, then it's impossible to document it as an instance attribute, but instead it's always documented as 2 seperate methods?

**Comment by lsegal** [comment]
> So I assume this means that when you have an attribute whose getter and setter are not as simple as attr_accessor :foo, then it's impossible to document it as an instance attribute,

It's not impossible, you can follow the details mentioned above to get this working, and that's the intended purpose, just note the caveats in usage. My point regarding complex attributes was specifically regarding source code showing up-- it should be possible to mark complex methods as "attributes" if you really want to.

**Comment by paddor** [comment]
I see. Now it works. Thanks! (I had no clue there actually have to be 2 blank lines.)

Honestly, I didn't know what's best practice here. The explanation about an implementation that is "doing something" vs. one that is only accessing another a property and returning it was very helpful. I guess in my case, this means that the attribute's source code really isn't relevant. And anybody who wants to see its implementation can just check out the source code.

(I only played around with read-only. It is read/write.)

Thanks a lot for your help.
